### PR TITLE
Add tokenizer factory to support plugin custom tokenizer

### LINKED ISSUE #7252: Customize field separators
### Description

Previously, field separators were added as parameter of Subfield with https://github.com/facebookincubator/velox/pull/6014. The only usage of Subfield in Gluten is to create SubfieldFilters like [SubstraitToVeloxPlan.cpp#L718-L726](https://github.com/facebookincubator/velox/blob/main/velox/substrait/SubstraitToVeloxPlan.cpp#L718-L726). We found it is not enough to simply customize separators there to disable tokenizing field name by dot.
There are many usages of Subfield in Velox scan, including [HiveDataSource.cpp#L762](https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/HiveDataSource.cpp#L762), [TableHandle.cpp#L193-L194](https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/TableHandle.cpp#L193-L194) and [ScanSpec.cpp#L422](https://github.com/facebookincubator/velox/blob/main/velox/dwio/common/ScanSpec.cpp#L422) etc. We cannot control them so the default separators are still used.
To make it work, now we need to modify the default separators in a hard-coded way by changing `dot`  to `'\0'`. Do you think it is possible to add a flag in Velox [flags](https://github.com/facebookincubator/velox/blob/main/velox/flag_definitions/flags.cpp
), so people can change the default content of separators?

### PR DESCRIPTION
**Pull request by zhli1142015**
Fixes #7252

### CODE REVIEW

**Comment by zhli1142015** [comment]
@rui-mo  and @Yuhta , could you please help review this?

File: velox/type/tests/SubfieldTest.cpp [hunk 1]
```diff
@@ -20,12 +20,11 @@
...
 
 std::vector<std::unique_ptr<Subfield::PathElement>> tokenize(
-    const std::string& path,
-    const std::shared_ptr<Separators>& separators = Separators::get()) {
+    const std::string& path) {
   std::vector<std::unique_ptr<Subfield::PathElement>> elements;
-  Tokenizer tokenizer(path, separators);
-  while (tokenizer.hasNext()) {
-    elements.push_back(tokenizer.next());
+  auto tokenizer = Tokenizer::getInstance(path);
```
Comment (rui-mo) [inline velox/type/tests/SubfieldTest.cpp:25]: Maybe add a unit test to register a dummy tokenizer, and test its customized behavior.

**Review by rui-mo — COMMENTED** [review]
Thanks. The idea looks nice to me. @Yuhta How do you think?

File: velox/type/tests/SubfieldTest.cpp (hunk 1 above)
Comment (zhli1142015) [reply velox/type/tests/SubfieldTest.cpp:25]: Sure, let me add one, thanks.

Comment (zhli1142015) [reply velox/type/tests/SubfieldTest.cpp:25]: Thanks, added.

File: velox/type/tests/SubfieldTest.cpp [hunk 2]
```diff
@@ -154,3 +148,52 @@ TEST(SubfieldTest, longSubscript) {
...
+ public:
+  explicit FakeTokenizer(const std::string& path) : path_(path) {
+    state = State::kNotReady;
+  }
+
+  bool hasNext() override {
+    switch (state) {
+      case State::kDone:
+        return false;
+      case State::kReady:
```
Comment (rui-mo) [inline velox/type/tests/SubfieldTest.cpp:162]: Seems only kDone and kNotReady are needed.

File: velox/type/tests/SubfieldTest.cpp [hunk 3]
```diff
@@ -154,3 +148,52 @@ TEST(SubfieldTest, longSubscript) {
...
+        break;
+      case State::kFailed:
+        VELOX_FAIL("Illegal state");
+    }
+    return true;
+  }
+
+  std::unique_ptr<Subfield::PathElement> next() override {
+    if (!hasNext()) {
+      VELOX_FAIL("No more tokens");
```
Comment (rui-mo) [inline velox/type/tests/SubfieldTest.cpp:174]: Nit: maybe use VELOX_USER_FAIL

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/type/tests/SubfieldTest.cpp (hunk 3 above)
Comment (zhli1142015) [reply velox/type/tests/SubfieldTest.cpp]: Updated, thanks.

File: velox/type/tests/SubfieldTest.cpp (hunk 2 above)
Comment (zhli1142015) [reply velox/type/tests/SubfieldTest.cpp]: Updated, thanks.

**Comment by zhli1142015** [comment]
Hello @Yuhta , could you please help to review?

**Comment by rui-mo** [comment]
We still need this PR. @Yuhta Gentel ping for review, thanks.

**Comment by rui-mo** [comment]
I believe we still need this PR (to disable the stale label).

**Comment by zhli1142015** [comment]
Thanks. let me rebase this PR.

**Comment by FelixYBW** [comment]
@zhli1142015 @rui-mo The PR is picked by oap/velox. Can you follow up?

**Comment by zhli1142015** [comment]
> @zhli1142015 @rui-mo The PR is picked by oap/velox. Can you follow up?

This is done long time ago, what I can remember is in Gluten we don't support to push down subfield filters to scan ops. And in this PR we add the interface to allow upstream to define the behavior self, but looks this is not reviewed from Meta ones.

**Comment by Yuhta** [comment]
@zhli1142015 @FelixYBW Velox is taking `Subfield` object directly in `HiveTableHandle`, so we don't need to add custom parser for it here.  You can add this parser in Gluten and pass the generated `Subfield` directly into Velox.  The default parser is specific to Presto and it's just kept in Velox for historical and convenience of unit testing reason.

**Comment by rui-mo** [comment]
> You can add this parser in Gluten and pass the generated Subfield directly into Velox.

@Yuhta Thanks for your review. In this manner, we are limited to control the Subfields generated by Gluten, while Velox also generates Subfields from string where incompatible separators for Spark are used. With this PR, we can customize all Subfields made by Velox and Gluten. What are your thoughts? Thanks!

**Comment by Yuhta** [comment]
@rui-mo I think the conversion from string to `Subfield` is one directional (at least in production code), no?

**Comment by rui-mo** [comment]
> I think the conversion from string to `Subfield` is one directional

@Yuhta That's right, while some Subfield created and used in Velox production code cannot be passed from Gluten to Velox. E.g,
https://github.com/facebookincubator/velox/blob/97ad9adb4f0ca3e781f2ac1ef0e0db3054c86904/velox/connectors/hive/HiveConnectorUtil.cpp#L419

https://github.com/facebookincubator/velox/blob/97ad9adb4f0ca3e781f2ac1ef0e0db3054c86904/velox/dwio/common/ScanSpec.cpp#L372

Please kindly correct me if I'm missing something. Thanks.

**Comment by Yuhta** [comment]
@rui-mo These are supposed to be single level field accessing and does not involve any parsing.  We can add another method `ScanSpec::getOrCreateChild(const std::string& name)` to replace these usage.  Would you create a PR for that?

**Comment by rui-mo** [comment]
> @rui-mo These are supposed to be single level field accessing and does not involve any parsing. We can add another method `ScanSpec::getOrCreateChild(const std::string& name)` to replace these usage. Would you create a PR for that?

@Yuhta Thanks for the suggestion. I will take a look.

**Pull request closed without merging**